salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,430 @@
+        return self.get('spec.rawBlockDevice.devicePath')
+
+    def format(self, force=False):
+        # We format an entire device, not just a partition: we need force=True.
+        super(RawBlockDevice, self).format(force=True)
+
+
+# }}}
+# Helpers {{{
+
+
+def _get_volume(name):
+    """Get a Volume object from the pillar."""
+    volume = __pillar__['metalk8s']['volumes'].get(name)
+    if volume is None:
+        raise ValueError('volume {} not found in pillar'.format(name))
+    if 'rawBlockDevice' in volume['spec']:
+        return RawBlockDevice(volume)
+    elif 'sparseLoopDevice' in volume['spec']:
+        return SparseLoopDevice(volume)
+    else:
+        raise ValueError('unsupported Volume type')
+
+
+def _run_cmd(cmd):
+    """Execute the given `cmd` command and return its result.
+
+    Raise `CommandExecutionError` if the command failed.
+
+    Args:
+        cmd  (str): command to execute
+
+    Returns:
+        dict: the command result (stderr, stdout, retcode, …)
+    """
+    ret = __salt__['cmd.run_all'](cmd)
+    if ret.get('retcode', 0) != 0:
+        raise CommandExecutionError(
+            'error while trying to run `{0}`: {1}' .format(cmd, ret['stderr'])
+        )
+    return ret
+
+
+def _quantity_to_bytes(quantity):
+    """Return a quantity with a unit converted into a number of bytes.
+
+    Examples:
+    >>> quantity_to_bytes('42Gi')
+    45097156608
+    >>> quantity_to_bytes('100M')
+    100000000
+    >>> quantity_to_bytes('1024')
+    1024
+
+    Args:
+        quantity (str): a quantity, composed of a count and an optional unit
+
+    Returns:
+        int: the capacity (in bytes)
+    """
+    UNIT_FACTOR = {
+      None:  1,
+      'Ki':  2 ** 10,
+      'Mi':  2 ** 20,
+      'Gi':  2 ** 30,
+      'Ti':  2 ** 40,
+      'Pi':  2 ** 50,
+      'k':  10 ** 3,
+      'M':  10 ** 6,
+      'G':  10 ** 9,
+      'T':  10 ** 12,
+      'P':  10 ** 15,
+    }
+    size_regex = r'^(?P<size>[1-9][0-9]*)(?P<unit>[kKMGTP]i?)?$'
+    match = re.match(size_regex, quantity)
+    assert match is not None, 'invalid resource.Quantity value'
+    size = int(match.groupdict()['size'])
+    unit = match.groupdict().get('unit')
+    return size * UNIT_FACTOR[unit]
+
+
+@contextlib.contextmanager
+def _open_fd(*args, **kwargs):
+    fd = os.open(*args, **kwargs)
+    try:
+        yield fd
+    finally:
+        os.close(fd)
+
+
+# Cannot use `disk.fstype` from Salt because you can't trust its output…
+#
+# Example:
+# - On a unformated sparse file:
+#     __salt__['disk.fstype']('/var/lib/metalk8s/storage/sparse/example')
+#     returns "/dev/sda1      ext4  41152736 7765776  31273476  20% /"
+#
+# - On a raw (non-formatted) block device:
+#     __salt__['disk.fstype']('/dev/sdb')
+#     returns "devtmpfs       devtmpfs   1932084     0   1932084   0% /dev"
+#
+# So yeah, let's not rely on this…
+def _get_fstype(path):
+    try:
+        res = _run_cmd('blkid {}'.format(path))
+        pat = r'TYPE="(?P<fstype>.+)"'
```

Comment:
Instead of using `blkid` and a regex to parse its output, consider using `lsblk`:
```
$ lsblk -o FSTYPE -r -n /dev/sda2
ext4
```
for actual block devices. Sadly enough, this seems not to work for files:
```
$ truncate -s 1M volume.dat
$ lsblk -o FSTYPE -r -n ./volume.dat
lsblk: ./volume.dat: not a block device
```

Reply:
The output of `lsblk` seems to be variable as well:
```
# lsblk -o FSTYPE -r -n /dev/sda1
ext4
# lsblk -o FSTYPE -r -n /dev/sda  # Why do I have an empty line in that case? oO

ext4
# lsblk -o FSTYPE -r -n /dev/sdb

```
And yeah, it doesn't work on files (even after the formatting it), so I'll stay on `blkid` I think (since it's already installed anyway), otherwise we can't check for the sparse file.